core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -186,8 +186,12 @@ class Partition(val topic: String,
       // record the epoch of the controller that made the leadership decision. This is useful while updating the isr
       // to maintain the decision maker controller's epoch in the zookeeper path
       controllerEpoch = partitionStateInfo.controllerEpoch
-      // add replicas that are new
-      allReplicas.foreach(replica => getOrCreateReplica(replica))
+      // add replicas that are new and reset lastCatchUpTime of non-isr replicas
+      allReplicas.foreach(replica => {
+        val r = getOrCreateReplica(replica)
+        if (!partitionStateInfo.isr.contains(replica))
+          r.resetLastCatchUpTime()
+      })
```

Comment:
@junrao There is another optimization I will make after your suggestion. That is, when broker receives LeaderAndIsrRequest to become leader of a partition, for each follower of this partition, the lastFetchTimeMs will be set to curTime and lastFetchLeaderLogEndOffset will be set to current LEO.